File: RR.cpp
#include <iostream>
#include <queue>
using namespace std;

class Process {
public:
    int pid, arrival, burst, remaining, waiting, tat, completion, finish;
};

int main() {
    int n, quantum;
    cout << "*** RR Scheduling (Preemptive) ***\n";

    cout << "Enter number of processes: ";  
    cin >> n;

    Process p[n];

    int sumBurst = 0;
    for(int i = 0; i < n; i++) {
        p[i].pid = i + 1;
        cout << "Enter arrival time for P" << p[i].pid << ": ";
        cin >> p[i].arrival;
        cout << "Enter burst time for P" << p[i].pid << ": ";
        cin >> p[i].burst;

        p[i].remaining = p[i].burst;
        p[i].finish = 0;
        sumBurst += p[i].burst;
        cout << endl;
    }

    cout << "\nEnter time quantum: ";
    cin >> quantum;

    // Sort by arrival time
    for(int i = 0; i < n - 1; i++) {
        for(int j = i + 1; j < n; j++) {
            if(p[i].arrival > p[j].arrival) {
                Process temp = p[i];
                p[i] = p[j];
                p[j] = temp;
            }
        }
    }

    queue<int> q;
    int time = p[0].arrival;
    q.push(0);

    while(!q.empty()) {
        int i = q.front();
        q.pop();

        if(p[i].remaining <= quantum) {
            time += p[i].remaining;
            p[i].remaining = 0;
            p[i].finish = 1;
            p[i].completion = time;
            p[i].tat = p[i].completion - p[i].arrival;
            p[i].waiting = p[i].tat - p[i].burst;
        }
        else {
            time += quantum;
            p[i].remaining -= quantum;
        }

        for(int j = 0; j < n; j++) {
            if(p[j].arrival <= time && p[j].finish == 0) {
                bool already = false;
                queue<int> temp = q;
                while(!temp.empty()) {
                    if(temp.front() == j) already = true;
                    temp.pop();
                }
                if(!already && j != i)
                    q.push(j);
            }
        }

        if(p[i].finish == 0)
            q.push(i);
    }

    cout << "\nProcessor\tArrival\tBurst\tCompletion\tTAT\tWaiting\n";
    cout << "-------------------------------------------------------------\n";

    double avgTAT = 0, avgWT = 0;

    for(int i = 0; i < n; i++) {
        cout << "P" << p[i].pid << "\t\t" << p[i].arrival << "\t" << p[i].burst 
             << "\t" << p[i].completion << "\t\t" << p[i].tat 
             << "\t" << p[i].waiting << endl;

        avgTAT += p[i].tat;
        avgWT += p[i].waiting;
    }

    cout << "\nAverage Turn Around Time: " << avgTAT / n;
    cout << "\nAverage Waiting Time: " << avgWT / n << endl;

    return 0;
}

File: RR.py
def round_robin(processes, tq):
    n = len(processes)
    at = [p[1] for p in processes]
    bt = [p[2] for p in processes]
    rt = bt[:]
    ct = [0] * n
    tat = [0] * n
    wt = [0] * n

    curr = 0
    completed = 0
    q = []
    in_queue = [False] * n
    gantt = []

    # Start from earliest arrival
    curr = min(at)

    while completed < n:
        # Add all processes that have arrived till now
        for i in range(n):
            if not in_queue[i] and at[i] <= curr and rt[i] > 0:
                q.append(i)
                in_queue[i] = True

        # If no process is ready, jump to next arrival
        if not q:
            next_arrival = min([at[i] for i in range(n) if rt[i] > 0], default=None)
            if next_arrival is None:
                break  # all done
            curr = max(curr + 1, next_arrival)
            continue

        # Pick first from queue
        i = q.pop(0)
        in_queue[i] = False

        exec_time = min(tq, rt[i])
        start = curr
        curr += exec_time
        rt[i] -= exec_time
        gantt.append((processes[i][0], start, curr))

        # Add new arrivals during execution
        for j in range(n):
            if not in_queue[j] and at[j] <= curr and rt[j] > 0:
                q.append(j)
                in_queue[j] = True

        # If process finished
        if rt[i] == 0:
            ct[i] = curr
            completed += 1
        else:
            q.append(i)
            in_queue[i] = True

    # Calculate TAT & WT
    for i in range(n):
        tat[i] = ct[i] - at[i]
        wt[i] = tat[i] - bt[i]

    # Gantt chart
    print("\nGantt Chart:")
    for g in gantt:
        print(f"| {g[0]} ", end="")
    print("|")
    print("0", end="")
    for g in gantt:
        print(f"   {g[2]}", end="")
    print("\n")

    # Table
    print("Process\tAT\tBT\tCT\tTAT\tWT")
    for i in range(n):
        print(f"{processes[i][0]}\t{at[i]}\t{bt[i]}\t{ct[i]}\t{tat[i]}\t{wt[i]}")

    avg_tat = sum(tat) / n
    avg_wt = sum(wt) / n
    print(f"\nAverage Turnaround Time = {avg_tat:.2f}")
    print(f"Average Waiting Time = {avg_wt:.2f}")


# ---------------- MAIN ----------------
if __name__ == "__main__":
    n = int(input("Enter number of processes: "))
    processes = []
    for i in range(n):
        pid = "P" + str(i + 1)
        at = int(input(f"Enter Arrival Time for {pid}: "))
        bt = int(input(f"Enter Burst Time for {pid}: "))
        processes.append((pid, at, bt))

    tq = int(input("Enter Time Quantum: "))
    round_robin(processes, tq)
